storage: bypass cached liveness in decommissioning, draining

tbg opened the pull request:
The liveness subsystem holds an in-memory map of "latest" livenesses
seen by heartbeats. This map is notoriously out-of-date and has a long
history of older information clobbering newer one (now dealt with to the
point where the introduction of sequence numbers on liveness records has
been deferred).

However, TestNodeLivenessDecommissionAbsent hits a race that will always
be present when a cache is used: it decommissions and recommissions a
node in quick succession from multiple nodes. Since the cache is
populated from Gossip, and these updates may reorder or just not have
arrived by the time they are needed, the test sometimes failed to update
the liveness record (since the cache terminated the request early based
on stale information):

1. node 1 sets node X to decommissioning.
2. node 2 starts to set node X to recommissioning
3. but when this executes node 2 doesn't yet know about the decommission
   in the first place,
4. so the cache thinks we're just doing a no-op
5. and the request becomes a no-op.
6. the node is still decommissioning, which is a problem.

We now bypass the cache for decommissioning-related operations (and
draining as well, for good measure). The cache is important only for
operations that run frequently, which is not the case for these.

Fixes #25029.
Fixes #24986.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25074)
<!-- Reviewable:end -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---

*[pkg/storage/node_liveness.go, line 352 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25074#-LAy5T4R6Up3rRfZzJFh:-LAy5T4R6Up3rRfZzJFi:bq1lhkg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4364f8c0deab154567fb3468b16126743c0e956e/pkg/storage/node_liveness.go#L352)):*
> ```Go
> type livenessUpdate struct {
> 	Liveness
> 	// when noCache is set, we won't assume that our in-memory cached version
> ```

Capital "w"

---

*[pkg/storage/node_liveness.go, line 356 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25074#-LAy5qpHD9UyYCSnT60A:-LAy5qpHD9UyYCSnT60B:b-aixyhz) ([raw file](https://github.com/cockroachdb/cockroach/blob/4364f8c0deab154567fb3468b16126743c0e956e/pkg/storage/node_liveness.go#L356)):*
> ```Go
> 	// table with whatever the client supplied. This is used for operations
> 	// that don't want to deal with the inconsistencies of using the cache.
> 	noCache bool
> ```

I have a slight preference for "ignoreCache" over "noCache", only because the former can't be mistaken as meaning "don't cache the result of this update". Your call.

---

*[pkg/storage/node_liveness.go, line 766 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25074#-LAy67YTDNfF1Xy7yemz:-LAy67YTDNfF1Xy7yen-:b-l2f0qb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4364f8c0deab154567fb3468b16126743c0e956e/pkg/storage/node_liveness.go#L766)):*
> ```Go
> 	// First check the existing liveness map to avoid known conditional
> 	// put failures.
> 	if l, err := nl.GetLiveness(update.NodeID); !update.noCache && err == nil &&
> ```

Let's avoid calling `GetLiveness` entirely when `!update.noCache`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25074#-:-LAy6KksE20Ep1ydsJza:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR!

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/storage/node_liveness.go, line 352 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25074#-LAy5T4R6Up3rRfZzJFh:-LB0_-kMByWC0BlJ-dqG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4364f8c0deab154567fb3468b16126743c0e956e/pkg/storage/node_liveness.go#L352)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Capital "w"
</blockquote></details>

Done.

---

*[pkg/storage/node_liveness.go, line 356 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25074#-LAy5qpHD9UyYCSnT60A:-LB0ZycC36v2KLquDjx9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4364f8c0deab154567fb3468b16126743c0e956e/pkg/storage/node_liveness.go#L356)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I have a slight preference for "ignoreCache" over "noCache", only because the former can't be mistaken as meaning "don't cache the result of this update". Your call.
</blockquote></details>

Done.

---

*[pkg/storage/node_liveness.go, line 766 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25074#-LAy67YTDNfF1Xy7yemz:-LB0_0pK7KnprT32JOhb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4364f8c0deab154567fb3468b16126743c0e956e/pkg/storage/node_liveness.go#L766)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Let's avoid calling `GetLiveness` entirely when `!update.noCache`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25074#-:-LB0_1mv70gpxYvv279H:b-814g8k)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: bypass cached liveness in decommissioning, draining

The liveness subsystem holds an in-memory map of "latest" livenesses
seen by heartbeats. This map is notoriously out-of-date and has a long
history of older information clobbering newer one (now dealt with to the
point where the introduction of sequence numbers on liveness records has
been deferred).

However, TestNodeLivenessDecommissionAbsent hits a race that will always
be present when a cache is used: it decommissions and recommissions a
node in quick succession from multiple nodes. Since the cache is
populated from Gossip, and these updates may reorder or just not have
arrived by the time they are needed, the test sometimes failed to update
the liveness record (since the cache terminated the request early based
on stale information):

1. node 1 sets node X to decommissioning.
2. node 2 starts to set node X to recommissioning
3. but when this executes node 2 doesn't yet know about the decommission
   in the first place,
4. so the cache thinks we're just doing a no-op
5. and the request becomes a no-op.
6. the node is still decommissioning, which is a problem.

We now bypass the cache for decommissioning-related operations (and
draining as well, for good measure). The cache is important only for
operations that run frequently, which is not the case for these.

Fixes #25029.
Fixes #24986.

Release note: None

tschottdorf commented:
bors r=nvanbenschoten
